localstack/services/s3/provider_native.py

```diff
@@ -0,0 +1,1353 @@
+    ) -> PutObjectOutput:
+        # TODO: validate order of validation
+        # TODO: still need to handle following parameters:
+        #  acl: ObjectCannedACL = None,
+        #  grant_full_control: GrantFullControl = None,
+        #  grant_read: GrantRead = None,
+        #  grant_read_acp: GrantReadACP = None,
+        #  grant_write_acp: GrantWriteACP = None,
+        #  -
+        #  request_payer: RequestPayer = None,
+        #  tagging: TaggingHeader = None,
+        #  object_lock_mode: ObjectLockMode = None,
+        #  object_lock_retain_until_date: ObjectLockRetainUntilDate = None,
+        #  object_lock_legal_hold_status: ObjectLockLegalHoldStatus = None,
+        store = self.get_store(context.account_id, context.region)
+        bucket_name = request["Bucket"]
+        if not (s3_bucket := store.buckets.get(bucket_name)):
+            raise NoSuchBucket("The specified bucket does not exist", BucketName=bucket_name)
+
+        if (
+            storage_class := request.get("StorageClass")
+        ) is not None and storage_class not in STORAGE_CLASSES:
+            raise InvalidStorageClass(
+                "The storage class you specified is not valid", StorageClassRequested=storage_class
+            )
+
+        if not config.S3_SKIP_KMS_KEY_VALIDATION and (sse_kms_key_id := request.get("SSEKMSKeyId")):
+            validate_kms_key_id(sse_kms_key_id, s3_bucket)
+
+        key = request["Key"]
+
+        system_metadata = get_system_metadata_from_request(request)
+        if not system_metadata.get("ContentType"):
+            system_metadata["ContentType"] = "binary/octet-stream"
+
+        # TODO: get all default from bucket once it is implemented
+        # validate encryption values
+
+        body = request.get("Body")
+        # check if chunked request
+        headers = context.request.headers
+        is_aws_chunked = headers.get("x-amz-content-sha256", "").startswith("STREAMING-")
+        if is_aws_chunked:
+            decoded_content_length = int(headers.get("x-amz-decoded-content-length", 0))
+            body = AwsChunkedDecoder(body, decoded_content_length)
+
+        # TODO check if key already exist, and if it is locked? so we don't override it
+
+        version_id = generate_version_id(s3_bucket.versioning_status)
+
+        fileobj = self._storage_backend.get_key_fileobj(
+            bucket_name=bucket_name,
+            object_key=key,
+            version_id=version_id,
+        )
+        # TODO: validate the algo?
+        checksum_value = None
+        if checksum_algorithm := request.get("ChecksumAlgorithm"):
+            if is_aws_chunked:
+                checksum_value = body.trailing_headers.get(
+                    f"x-amz-checksum-{checksum_algorithm.lower()}"
+                )
+            else:
+                checksum_value = request.get(f"Checksum{checksum_algorithm.upper()}")
+
+        size, etag, calculated_checksum_value = readinto_fileobj(body, fileobj, checksum_algorithm)
```

Comment:
Q: Shouldn't, conceptually, this entire logic be encapsulated in the storage backend? it seems to me that the decision when and how to write the object into a file object is the responsibility of the storage backend, not the provider.

as a thought experiment, i'm thinking about how to implement a "no-op" storage backend that just discards the files.

Reply:
I think we could implement `readinto_fileobj` inside the `StorageBackend`, but it would basically one operation for each different provider call, I think one method for each provider method: a different one for `CopyObject`, another one for `UploadPart`, `UploadCopyPart` and last for `PutObject`. Then we're coupling `StoragBackend` to the provider a lot. Not sure how much decoupling we really need